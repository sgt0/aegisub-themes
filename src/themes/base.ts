export type AegisubTheme = {
  name: string;
  subtitle: {
    background: string;
    syntax: {
      brackets: string;
      comments: string;
      drawingCommands: string;
      drawingXCoords: string;
      drawingYCoords: string;
      lineBreak: string;
      normal: string;
      parameters: string;

      /** Slashes and Parentheses. */
      slashes: string;
    };
  };
  subtitleGrid: {
    background: {
      background: string;
      closedFold: string;
      comment: string;
      inframe: string;
      openFold: string;
      selectedComment: string;
      selection: string;
    };

    /** Collision foreground. */
    collision: string;
    cpsError: string;
    header: string;
    leftColumn: string;
    lines: string;

    /** Selection foreground. */
    selection: string;

    /** Standard foreground. */
    standard: string;
  };
};
