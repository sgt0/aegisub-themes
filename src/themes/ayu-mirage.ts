import { fromBase16, mergeAegisubThemes, type Base16Palette } from './utils';

const PALETTE: Base16Palette = {
  base00: '#171B24',
  base01: '#1F2430',
  base02: '#242936',
  base03: '#707A8C',
  base04: '#8A9199',
  base05: '#CCCAC2',
  base06: '#D9D7CE',
  base07: '#F3F4F5',
  base08: '#F28779',
  base09: '#FFAD66',
  base0A: '#FFD173',
  base0B: '#D5FF80',
  base0C: '#95E6CB',
  base0D: '#5CCFE6',
  base0E: '#D4BFFF',
  base0F: '#F29E74'
};

export const AYU_MIRAGE_THEME = mergeAegisubThemes(
  { name: 'Ayu Mirage' },
  fromBase16(PALETTE),
  {
    subtitleGrid: {
      background: {
        closedFold: '#252c3e',
        openFold: '#252c3e',
        selectedComment: '#67532A',
        selection: '#67532A'
      },
      header: '#225660'
    }
  }
);
