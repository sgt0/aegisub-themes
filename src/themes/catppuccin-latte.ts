import { fromBase16, mergeAegisubThemes, type Base16Palette } from './utils';

const PALETTE: Base16Palette = {
  base00: '#eff1f5',
  base01: '#e6e9ef',
  base02: '#ccd0da',
  base03: '#bcc0cc',
  base04: '#acb0be',
  base05: '#4c4f69',
  base06: '#dc8a78',
  base07: '#7287fd',
  base08: '#d20f39',
  base09: '#fe640b',
  base0A: '#df8e1d',
  base0B: '#40a02b',
  base0C: '#179299',
  base0D: '#1e66f5',
  base0E: '#8839ef',
  base0F: '#dd7878'
};

export const CATPPUCCIN_LATTE_THEME = mergeAegisubThemes(
  { name: 'Catppuccin Latte' },
  fromBase16(PALETTE),
  {
    subtitleGrid: {
      background: {
        selectedComment: '#FCB0A1',
        selection: '#FCB0A1'
      },

      // No way to make base0D work contrast-wise. This alternative is derived
      // from base07.
      header: '#B7BFFB'
    }
  }
);
