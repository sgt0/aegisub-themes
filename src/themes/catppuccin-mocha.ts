import {
  fromBase16,
  meetContrastRatio,
  mergeAegisubThemes,
  type Base16Palette
} from './utils';

const PALETTE: Base16Palette = {
  base00: '#1e1e2e',
  base01: '#181825',
  base02: '#313244',
  base03: '#45475a',
  base04: '#585b70',
  base05: '#cdd6f4',
  base06: '#f5e0dc',
  base07: '#b4befe',
  base08: '#f38ba8',
  base09: '#fab387',
  base0A: '#f9e2af',
  base0B: '#a6e3a1',
  base0C: '#94e2d5',
  base0D: '#89b4fa',
  base0E: '#cba6f7',
  base0F: '#f2cdcd'
};

const selectionBg = meetContrastRatio(PALETTE.base05, PALETTE.base0B).hex();

export const CATPPUCCIN_MOCHA_THEME = mergeAegisubThemes(
  { name: 'Catppuccin Mocha' },
  fromBase16(PALETTE),
  {
    subtitleGrid: {
      background: {
        selectedComment: selectionBg,
        selection: selectionBg
      },
      header: meetContrastRatio(PALETTE.base05, PALETTE.base0D).hex()
    }
  }
);
