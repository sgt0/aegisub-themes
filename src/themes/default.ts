import type { AegisubTheme } from './base';

export const DEFAULT_THEME: AegisubTheme = {
  name: 'Default',
  subtitle: {
    background: 'rgb(255, 255, 255)',
    syntax: {
      brackets: 'rgb(20, 50, 255)',
      comments: 'rgb(0,0,0)',
      drawingCommands: 'rgb(0,0,0)',
      drawingXCoords: 'rgb(90,40,40)',
      drawingYCoords: 'rgb(40,90,40)',
      lineBreak: 'rgb(160, 160, 160)',
      normal: 'rgb(0,0,0)',
      parameters: 'rgb(40, 90, 40)',
      slashes: 'rgb(255, 0, 200)'
    }
  },
  subtitleGrid: {
    background: {
      background: 'rgb(255,255,255)',
      closedFold: 'rgb(200, 200, 200)',
      comment: 'rgb(216, 222, 245)',
      inframe: 'rgb(255, 253, 234)',
      openFold: 'rgb(235, 235, 235)',
      selectedComment: 'rgb(211, 238, 238)',
      selection: 'rgb(206, 255, 231)'
    },
    collision: 'rgb(255,0,0)',
    cpsError: 'rgb(255,0,0)',
    header: 'rgb(165, 207, 231)',
    leftColumn: 'rgb(196, 236, 201)',
    lines: 'rgb(190,190,190)',
    selection: 'rgb(0,0,0)',
    standard: 'rgb(0,0,0)'
  }
};
