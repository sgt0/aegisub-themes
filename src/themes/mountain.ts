import {
  fromBase16,
  meetContrastRatio,
  mergeAegisubThemes,
  type Base16Palette
} from './utils';

const PALETTE: Base16Palette = {
  base00: '#0f0f0f',
  base01: '#191919',
  base02: '#262626',
  base03: '#4c4c4c',
  base04: '#ac8a8c',
  base05: '#cacaca',
  base06: '#e7e7e7',
  base07: '#f0f0f0',
  base08: '#ac8a8c',
  base09: '#ceb188',
  base0A: '#aca98a',
  base0B: '#8aac8b',
  base0C: '#8aabac',
  base0D: '#8f8aac',
  base0E: '#ac8aac',
  base0F: '#ac8a8c'
};

const selectionBg = meetContrastRatio(PALETTE.base05, PALETTE.base0B).hex();

export const MOUNTAIN_THEME = mergeAegisubThemes(
  { name: 'Mountain' },
  fromBase16(PALETTE),
  {
    subtitleGrid: {
      background: {
        comment: meetContrastRatio(PALETTE.base05, PALETTE.base03).hex(),
        inframe: meetContrastRatio(PALETTE.base05, PALETTE.base0E).hex(),
        selectedComment: selectionBg,
        selection: selectionBg
      },
      header: meetContrastRatio(PALETTE.base05, PALETTE.base0D).hex()
    }
  }
);
