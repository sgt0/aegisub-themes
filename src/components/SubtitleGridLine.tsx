import { css, useTheme } from '@emotion/react';
import chroma, { Color } from 'chroma-js';
import { AegisubTheme } from '../themes/base';

const CPS_MIN = 15;
const CPS_MAX = 30;

const styles = {
  minCell: css({ whiteSpace: 'nowrap', width: '1%' }),
  textAlignCenter: css({ textAlign: 'center' })
};

/** Blend two colors in the same way that wxWidgets does. */
function wxAlphaBlend(fg: number, bg: number, alpha: number): number {
  const result = bg + alpha * (fg - bg);
  return Math.min(Math.max(result, 0), 255);
}

/** Blend two colors in the same way that Aegisub does. */
function aegiBlend(fg: Color, bg: Color, alpha: number): Color {
  const fgRgb = fg.rgb(false);
  const bgRgb = bg.rgb(false);
  return chroma(
    wxAlphaBlend(fgRgb[0], bgRgb[0], alpha),
    wxAlphaBlend(fgRgb[1], bgRgb[1], alpha),
    wxAlphaBlend(fgRgb[2], bgRgb[2], alpha)
  );
}

type Props = React.DetailedHTMLProps<
  React.TableHTMLAttributes<HTMLTableRowElement>,
  HTMLTableRowElement
> & {
  cps: number;
  end: string;
  isComment?: boolean;
  isFold?: boolean;
  isInframe?: boolean;
  isSelected?: boolean;
  lineNumber: number;
  start: string;
  style: string;
  text: string;
};

export function SubtitleGridLine({
  cps,
  end,
  isComment = false,
  isFold = false,
  isInframe = false,
  isSelected = false,
  lineNumber,
  onClick,
  start,
  style,
  text
}: Props) {
  const theme: AegisubTheme = useTheme() as AegisubTheme;

  const backgroundColor = isSelected
    ? isComment
      ? theme.subtitleGrid.background.selectedComment
      : theme.subtitleGrid.background.selection
    : isComment
      ? theme.subtitleGrid.background.comment
      : isInframe
        ? theme.subtitleGrid.background.inframe
        : isFold
          ? theme.subtitleGrid.background.closedFold
          : theme.subtitleGrid.background.background;

  let cpsBackgroundColor = backgroundColor;
  let cpsColor = 'inherit';
  if (cps > CPS_MIN) {
    const alpha = Math.min((cps - CPS_MIN + 1) / (CPS_MAX - CPS_MIN + 1), 1);
    cpsBackgroundColor = aegiBlend(
      chroma(theme.subtitleGrid.cpsError),
      chroma(backgroundColor),
      alpha
    ).hex();
    cpsColor = aegiBlend(
      chroma('black'),
      chroma(theme.subtitleGrid.standard),
      alpha
    ).hex();
  }

  return (
    <tr css={css({ backgroundColor })} onClick={onClick}>
      <th
        css={css(styles.textAlignCenter, styles.minCell, {
          backgroundColor: theme.subtitleGrid.leftColumn
        })}
        scope="row"
      >
        {lineNumber}
      </th>
      <td css={css(styles.textAlignCenter, styles.minCell)}>{isFold && '>'}</td>
      <td css={css(styles.textAlignCenter, styles.minCell)}>{start}</td>
      <td css={css(styles.textAlignCenter, styles.minCell)}>{end}</td>
      <td
        css={css(styles.textAlignCenter, styles.minCell, {
          backgroundColor: cpsBackgroundColor,
          color: cpsColor
        })}
      >
        {cps}
      </td>
      <td css={css(styles.minCell)}>{style}</td>
      <td>{text}</td>
    </tr>
  );
}
