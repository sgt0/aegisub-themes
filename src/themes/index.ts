import { AYU_DARK_THEME } from './ayu-dark';
import { AYU_LIGHT_THEME } from './ayu-light';
import { AYU_MIRAGE_THEME } from './ayu-mirage';
import { CATPPUCCIN_LATTE_THEME } from './catppuccin-latte';
import { CATPPUCCIN_MOCHA_THEME } from './catppuccin-mocha';
import { DEFAULT_THEME } from './default';
import { DRACULA_THEME } from './dracula';
import { MOUNTAIN_THEME } from './mountain';
import { POPPO_THEME } from './poppo';
import { SAKURA_THEME } from './sakura';
import { SETI_THEME } from './seti';
import { SOLARIZED_DARK_THEME } from './solarized-dark';
import { SOLARIZED_LIGHT_THEME } from './solarized-light';
import { ZENBURN_THEME } from './zenburn';

export const THEMES = [
  AYU_DARK_THEME,
  AYU_LIGHT_THEME,
  AYU_MIRAGE_THEME,
  CATPPUCCIN_LATTE_THEME,
  CATPPUCCIN_MOCHA_THEME,
  DEFAULT_THEME,
  DRACULA_THEME,
  MOUNTAIN_THEME,
  POPPO_THEME,
  SAKURA_THEME,
  SETI_THEME,
  SOLARIZED_DARK_THEME,
  SOLARIZED_LIGHT_THEME,
  ZENBURN_THEME
];
