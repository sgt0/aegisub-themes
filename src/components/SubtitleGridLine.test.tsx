// @vitest-environment happy-dom

import { createSerializer } from '@emotion/jest';
import { ThemeProvider } from '@emotion/react';
import { render } from '@testing-library/react';
import { expect, test } from 'vitest';
import { DEFAULT_THEME } from '../themes/default';
import { SubtitleGridLine } from './SubtitleGridLine';

expect.addSnapshotSerializer(createSerializer());

test('comment line', () => {
  expect(
    render(
      <ThemeProvider theme={DEFAULT_THEME}>
        <table>
          <tbody>
            <SubtitleGridLine
              cps={6}
              end={'0:00:08.00'}
              isComment={true}
              lineNumber={1}
              start={'0:00:05.00'}
              style={'Default'}
              text={'This is a comment line'}
            />
          </tbody>
        </table>
      </ThemeProvider>
    ).asFragment()
  ).toMatchSnapshot();
});
