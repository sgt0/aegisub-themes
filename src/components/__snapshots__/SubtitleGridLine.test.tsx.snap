// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`comment line 1`] = `
<DocumentFragment>
  .emotion-0 {
  background-color: rgb(216, 222, 245);
}

.emotion-1 {
  text-align: center;
  white-space: nowrap;
  width: 1%;
  background-color: rgb(196, 236, 201);
}

.emotion-2 {
  text-align: center;
  white-space: nowrap;
  width: 1%;
}

.emotion-5 {
  text-align: center;
  white-space: nowrap;
  width: 1%;
  background-color: rgb(216, 222, 245);
  color: inherit;
}

.emotion-6 {
  white-space: nowrap;
  width: 1%;
}

<table>
    <tbody>
      <tr
        class="emotion-0"
      >
        <th
          class="emotion-1"
          scope="row"
        >
          1
        </th>
        <td
          class="emotion-2"
        />
        <td
          class="emotion-2"
        >
          0:00:05.00
        </td>
        <td
          class="emotion-2"
        >
          0:00:08.00
        </td>
        <td
          class="emotion-5"
        >
          6
        </td>
        <td
          class="emotion-6"
        >
          Default
        </td>
        <td>
          This is a comment line
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
